File: app/src/test/java/org/alkaline/taskbrain/dsl/cache/DependencyAnalyzerTest.kt
package org.alkaline.taskbrain.dsl.cache

import org.alkaline.taskbrain.dsl.language.Lexer
import org.alkaline.taskbrain.dsl.language.Parser
import org.junit.Assert.*
import org.junit.Test

/**
 * Tests for DependencyAnalyzer.
 * Phase 2: AST analysis for dependency detection.
 */
class DependencyAnalyzerTest {

    // region Field access detection - path

    @Test
    fun `detects path dependency from property access`() {
        val analysis = analyze("[.path]")
        assertTrue(analysis.dependsOnPath)
    }

    @Test
    fun `detects path dependency from find path argument`() {
        val analysis = analyze("[find(path: \"inbox\")]")
        assertTrue(analysis.dependsOnPath)
    }

    @Test
    fun `no path dependency for unrelated code`() {
        val analysis = analyze("[42]")
        assertFalse(analysis.dependsOnPath)
    }

    // endregion

    // region Field access detection - modified

    @Test
    fun `detects modified dependency from property access`() {
        val analysis = analyze("[.modified]")
        assertTrue(analysis.dependsOnModified)
    }

    @Test
    fun `detects modified in where clause`() {
        val analysis = analyze("[find(where: [i.modified.gt(\"2026-01-01\")])]")
        assertTrue(analysis.dependsOnModified)
    }

    // endregion

    // region Field access detection - created

    @Test
    fun `detects created dependency from property access`() {
        val analysis = analyze("[.created]")
        assertTrue(analysis.dependsOnCreated)
    }

    // endregion

    // region Content access detection - name

    @Test
    fun `detects first line access from name property`() {
        val analysis = analyze("[.name]")
        assertTrue(analysis.accessesFirstLine)
        assertFalse(analysis.accessesNonFirstLine)
    }

    @Test
    fun `detects first line access from find name argument`() {
        val analysis = analyze("[find(name: \"Shopping\")]")
        assertTrue(analysis.accessesFirstLine)
    }

    // endregion

    // region Content access detection - content

    @Test
    fun `detects full content access from content property`() {
        val analysis = analyze("[.content]")
        assertTrue(analysis.accessesFirstLine)
        assertTrue(analysis.accessesNonFirstLine)
    }

    // endregion

    // region find() detection

    @Test
    fun `detects note existence dependency from find`() {
        val analysis = analyze("[find(path: \"inbox\")]")
        assertTrue(analysis.dependsOnNoteExistence)
    }

    @Test
    fun `detects find with where clause`() {
        val analysis = analyze("[find(where: [i.path.startsWith(\"inbox\")])]")
        assertTrue(analysis.dependsOnNoteExistence)
        assertTrue(analysis.dependsOnPath) // from i.path access
    }

    @Test
    fun `no note existence dependency without find`() {
        val analysis = analyze("[.path]")
        assertFalse(analysis.dependsOnNoteExistence)
    }

    // endregion

    // region Hierarchy access detection - .up

    @Test
    fun `detects up hierarchy access`() {
        val analysis = analyze("[.up]")
        assertEquals(1, analysis.hierarchyAccesses.size)
        assertEquals(HierarchyPath.Up, analysis.hierarchyAccesses[0].path)
        assertNull(analysis.hierarchyAccesses[0].field)
    }

    @Test
    fun `detects up method call`() {
        val analysis = analyze("[.up()]")
        assertEquals(1, analysis.hierarchyAccesses.size)
        assertEquals(HierarchyPath.Up, analysis.hierarchyAccesses[0].path)
    }

    @Test
    fun `detects up with levels`() {
        val analysis = analyze("[.up(2)]")
        assertEquals(1, analysis.hierarchyAccesses.size)
        val path = analysis.hierarchyAccesses[0].path
        assertTrue(path is HierarchyPath.UpN)
        assertEquals(2, (path as HierarchyPath.UpN).levels)
    }

    @Test
    fun `detects up with field access`() {
        val analysis = analyze("[.up.path]")
        assertTrue(analysis.dependsOnPath)
        assertEquals(1, analysis.hierarchyAccesses.size)
        assertEquals(HierarchyPath.Up, analysis.hierarchyAccesses[0].path)
        assertEquals(NoteField.PATH, analysis.hierarchyAccesses[0].field)
    }

    @Test
    fun `detects up with name access`() {
        val analysis = analyze("[.up.name]")
        assertTrue(analysis.accessesFirstLine)
        assertEquals(1, analysis.hierarchyAccesses.size)
        assertEquals(NoteField.NAME, analysis.hierarchyAccesses[0].field)
    }

    @Test
    fun `detects chained up access`() {
        val analysis = analyze("[.up.up]")
        // Should have one access with UpN(2)
        assertEquals(1, analysis.hierarchyAccesses.size)
        val path = analysis.hierarchyAccesses[0].path
        assertTrue(path is HierarchyPath.UpN)
        assertEquals(2, (path as HierarchyPath.UpN).levels)
    }

    // endregion

    // region Hierarchy access detection - .root

    @Test
    fun `detects root hierarchy access`() {
        val analysis = analyze("[.root]")
        assertEquals(1, analysis.hierarchyAccesses.size)
        assertEquals(HierarchyPath.Root, analysis.hierarchyAccesses[0].path)
    }

    @Test
    fun `detects root with field access`() {
        val analysis = analyze("[.root.modified]")
        assertTrue(analysis.dependsOnModified)
        assertEquals(1, analysis.hierarchyAccesses.size)
        assertEquals(HierarchyPath.Root, analysis.hierarchyAccesses[0].path)
        assertEquals(NoteField.MODIFIED, analysis.hierarchyAccesses[0].field)
    }

    // endregion

    // region Statement lists

    @Test
    fun `analyzes all statements in statement list`() {
        val analysis = analyze("[a: .path; b: .modified; a]")
        assertTrue(analysis.dependsOnPath)
        assertTrue(analysis.dependsOnModified)
    }

    // endregion

    // region Lambda analysis

    @Test
    fun `analyzes lambda body`() {
        val analysis = analyze("[[i.path]]")
        assertTrue(analysis.dependsOnPath)
    }

    @Test
    fun `analyzes lambda invocation`() {
        val analysis = analyze("[f: [i.modified]; f(.)]")
        assertTrue(analysis.dependsOnModified)
    }

    // endregion

    // region Execution blocks

    @Test
    fun `analyzes once block body`() {
        val analysis = analyze("[once[.path]]")
        assertTrue(analysis.dependsOnPath)
    }

    @Test
    fun `analyzes refresh block body`() {
        val analysis = analyze("[refresh[.modified]]")
        assertTrue(analysis.dependsOnModified)
    }

    // endregion

    // region toPartialDependencies

    @Test
    fun `toPartialDependencies converts analysis to dependencies`() {
        val analysis = analyze("[find(path: \"inbox\"); .up.modified]")
        val deps = analysis.toPartialDependencies()

        assertTrue(deps.dependsOnPath)
        assertTrue(deps.dependsOnModified)
        assertTrue(deps.dependsOnNoteExistence)
        assertTrue(deps.firstLineNotes.isEmpty()) // Filled at runtime
        assertTrue(deps.hierarchyDeps.isEmpty())  // Filled at runtime
    }

    // endregion

    // region Complex expressions

    @Test
    fun `complex find with multiple dependencies`() {
        val analysis = analyze(
            "[find(path: pattern(digit*4), where: [i.modified.gt(\"2026-01-01\")])]"
        )
        assertTrue(analysis.dependsOnNoteExistence)
        assertTrue(analysis.dependsOnPath)
        assertTrue(analysis.dependsOnModified)
    }

    @Test
    fun `view inherits dependencies from input`() {
        val analysis = analyze("[view find(path: \"inbox\")]")
        assertTrue(analysis.dependsOnNoteExistence)
        assertTrue(analysis.dependsOnPath)
    }

    @Test
    fun `method call on find result`() {
        val analysis = analyze("[first(find(path: \"inbox\")).name]")
        assertTrue(analysis.dependsOnNoteExistence)
        assertTrue(analysis.dependsOnPath)
        assertTrue(analysis.accessesFirstLine)
    }

    // endregion

    // region DirectiveAnalysis.EMPTY

    @Test
    fun `EMPTY has no dependencies`() {
        val empty = DirectiveAnalysis.EMPTY
        assertFalse(empty.dependsOnPath)
        assertFalse(empty.dependsOnModified)
        assertFalse(empty.dependsOnCreated)
        assertFalse(empty.dependsOnNoteExistence)
        assertFalse(empty.accessesFirstLine)
        assertFalse(empty.accessesNonFirstLine)
        assertTrue(empty.hierarchyAccesses.isEmpty())
    }

    // endregion

    private fun analyze(code: String): DirectiveAnalysis {
        val tokens = Lexer(code).tokenize()
        val directive = Parser(tokens, code).parseDirective()
        return DependencyAnalyzer.analyze(directive.expression)
    }
}

File: web/src/__tests__/dsl/cache/DependencyAnalyzer.test.ts
import { describe, it, expect } from 'vitest'
import { analyze, analysisToPartialDependencies } from '../../../dsl/cache/DependencyAnalyzer'
import type { Expression } from '../../../dsl/language/Expression'

function selfRef(position = 0): Expression {
  return { kind: 'CurrentNoteRef', position }
}

function prop(target: Expression, property: string, position = 0): Expression {
  return { kind: 'PropertyAccess', target, property, position }
}

function call(name: string, args: Expression[] = [], namedArgs: { name: string; value: Expression; position: number }[] = [], position = 0): Expression {
  return { kind: 'CallExpr', name, args, namedArgs, position }
}

function methodCall(target: Expression, methodName: string, args: Expression[] = [], namedArgs: { name: string; value: Expression; position: number }[] = [], position = 0): Expression {
  return { kind: 'MethodCall', target, methodName, args, namedArgs, position }
}

function numLit(value: number, position = 0): Expression {
  return { kind: 'NumberLiteral', value, position }
}

function strLit(value: string, position = 0): Expression {
  return { kind: 'StringLiteral', value, position }
}

function varRef(name: string, position = 0): Expression {
  return { kind: 'VariableRef', name, position }
}

function lambda(params: string[], body: Expression, position = 0): Expression {
  return { kind: 'LambdaExpr', params, body, position }
}

describe('DependencyAnalyzer', () => {
  describe('field dependency detection', () => {
    it('detects path dependency', () => {
      const result = analyze(prop(selfRef(), 'path'))
      expect(result.dependsOnPath).toBe(true)
      expect(result.dependsOnModified).toBe(false)
    })

    it('detects modified dependency', () => {
      const result = analyze(prop(selfRef(), 'modified'))
      expect(result.dependsOnModified).toBe(true)
    })

    it('detects created dependency', () => {
      const result = analyze(prop(selfRef(), 'created'))
      expect(result.dependsOnCreated).toBe(true)
    })

    it('detects name access (first line)', () => {
      const result = analyze(prop(selfRef(), 'name'))
      expect(result.accessesFirstLine).toBe(true)
    })

    it('detects content access (first and non-first line)', () => {
      const result = analyze(prop(selfRef(), 'content'))
      expect(result.accessesFirstLine).toBe(true)
      expect(result.accessesNonFirstLine).toBe(true)
    })
  })

  describe('find predicates', () => {
    it('find with path depends on note existence and path', () => {
      const expr = call('find', [], [
        { name: 'path', value: strLit('a/b'), position: 0 },
      ])
      const result = analyze(expr)
      expect(result.dependsOnNoteExistence).toBe(true)
      expect(result.dependsOnPath).toBe(true)
    })

    it('find with name depends on note existence and all names', () => {
      const expr = call('find', [], [
        { name: 'name', value: strLit('test'), position: 0 },
      ])
      const result = analyze(expr)
      expect(result.dependsOnNoteExistence).toBe(true)
      expect(result.dependsOnAllNames).toBe(true)
      expect(result.accessesFirstLine).toBe(true)
    })

    it('find with where analyzes the where expression', () => {
      const whereBody = prop(varRef('note'), 'path')
      const expr = call('find', [], [
        { name: 'where', value: lambda(['note'], whereBody), position: 0 },
      ])
      const result = analyze(expr)
      expect(result.dependsOnNoteExistence).toBe(true)
    })
  })

  describe('mutation detection', () => {
    it('detects assignment as mutating', () => {
      const expr: Expression = {
        kind: 'Assignment',
        target: varRef('x'),
        value: numLit(1),
        position: 0,
      }
      const result = analyze(expr)
      expect(result.isMutating).toBe(true)
    })

    it('non-mutating expression is not flagged', () => {
      const result = analyze(prop(selfRef(), 'name'))
      expect(result.isMutating).toBe(false)
    })
  })

  describe('hierarchy access detection', () => {
    it('detects up property access on self', () => {
      const expr = prop(selfRef(), 'up')
      const result = analyze(expr)
      expect(result.hierarchyAccesses.length).toBeGreaterThan(0)
      expect(result.hierarchyAccesses[0]!.path.kind).toBe('Up')
    })

    it('detects root property access on self', () => {
      const expr = prop(selfRef(), 'root')
      const result = analyze(expr)
      expect(result.hierarchyAccesses.length).toBeGreaterThan(0)
      expect(result.hierarchyAccesses[0]!.path.kind).toBe('Root')
    })

    it('detects up method call with levels', () => {
      const expr = methodCall(selfRef(), 'up', [numLit(3)])
      const result = analyze(expr)
      expect(result.hierarchyAccesses.length).toBeGreaterThan(0)
      const access = result.hierarchyAccesses[0]!
      expect(access.path.kind).toBe('UpN')
      if (access.path.kind === 'UpN') {
        expect(access.path.levels).toBe(3)
      }
    })

    it('detects field access on hierarchy path', () => {
      // .up.name
      const expr = prop(prop(selfRef(), 'up'), 'name')
      const result = analyze(expr)
      const fieldAccesses = result.hierarchyAccesses.filter((a) => a.field !== null)
      expect(fieldAccesses.length).toBeGreaterThan(0)
    })

    it('detects chained up access', () => {
      // .up.up -> UpN(2)
      const expr = prop(prop(selfRef(), 'up'), 'up')
      const result = analyze(expr)
      const upAccesses = result.hierarchyAccesses.filter((a) => a.path.kind === 'UpN')
      expect(upAccesses.length).toBeGreaterThan(0)
      if (upAccesses[0]!.path.kind === 'UpN') {
        expect(upAccesses[0]!.path.levels).toBe(2)
      }
    })
  })

  describe('analysisToPartialDependencies', () => {
    it('converts analysis to dependencies', () => {
      const analysis = analyze(call('find', [], [
        { name: 'path', value: strLit('x'), position: 0 },
      ]))
      const deps = analysisToPartialDependencies(analysis)
      expect(deps.dependsOnPath).toBe(true)
      expect(deps.dependsOnNoteExistence).toBe(true)
    })
  })

  describe('statement list analysis', () => {
    it('combines dependencies from multiple statements', () => {
      const expr: Expression = {
        kind: 'StatementList',
        statements: [
          prop(selfRef(), 'path'),
          prop(selfRef(), 'modified'),
        ],
        position: 0,
      }
      const result = analyze(expr)
      expect(result.dependsOnPath).toBe(true)
      expect(result.dependsOnModified).toBe(true)
    })
  })

  describe('lambda analysis', () => {
    it('analyzes lambda body for dependencies', () => {
      const expr = lambda(['x'], prop(selfRef(), 'modified'))
      const result = analyze(expr)
      expect(result.dependsOnModified).toBe(true)
    })
  })
})
